### PR TITLE
Feature/adding nodelets

### PR DESCRIPTION
**Pull request by SGSSGene**
Work In Progress. Don't merge. Pending on PR #341 

adds nodelet for ekf, ukf and navsat_transform

Because Node and Nodelets handle NodeHandles a little different. I needed to move
NodeHandle and private NodeHandle from RosFilter and NavSatTransform into
the main function and the Nodelet implementation.

I added a test_[ekf|ukf]_localization_nodelet_bag1.test to show that
nodelets works just as nodes do.

Also an example is shown in ekf_nodelet_template.launch on how the
ekf_localization can be used with easy switch between node and nodelet
in a more complex environment.

### CODE REVIEW

File: include/robot_localization/ros_filter.h [hunk 1]
```diff
@@ -594,6 +600,41 @@ template<class T> class RosFilter
     // front() refers to the measurement with the earliest timestamp.
     // back() refers to the measurement with the latest timestamp.
     MeasurementHistoryDeque measurementHistory_;
+
+    //! @brief broadcaster of worldTransform tfs
+    //!
+    tf2_ros::TransformBroadcaster worldTransformBroadcaster_;
+
```
Comment (ayrton04) [inline include/robot_localization/ros_filter.h:626]: Extra space

File: include/robot_localization/ros_filter.h [hunk 2]
```diff
@@ -594,6 +600,41 @@ template<class T> class RosFilter
...
+    //! @brief last call of periodicUpdate
+    //!
+    ros::Time lastDiagTime_;
+
+    //! @brief timer calling periodicUpdate
+    //!
+    ros::Timer periodicUpdateTimer_;
+
+	//! @brief minimal frequency
```
Comment (ayrton04) [inline include/robot_localization/ros_filter.h]: Indentation

File: include/robot_localization/ros_filter.h [hunk 3]
```diff
@@ -594,6 +600,41 @@ template<class T> class RosFilter
...
+	//!
+    double minFrequency_;
+
+    //! @brief maximal frequency
+    double maxFrequency_;
+
```
Comment (ayrton04) [inline include/robot_localization/ros_filter.h]: Extra spaces. Might want to run roslint on the whole thing.

File: package.xml [hunk 4]
```diff
@@ -32,15 +32,18 @@
   <build_depend>message_generation</build_depend>
   <build_depend>python-catkin-pkg</build_depend>
   <build_depend>roslint</build_depend>
+  <build_depend>nodelet</build_depend>
```
Comment (ayrton04) [inline package.xml:36]: Please alphabetize

File: src/ekf_localization_node.cpp [hunk 5]
```diff
@@ -38,9 +38,12 @@ int main(int argc, char **argv)
...
+  ros::NodeHandle nh;
+  ros::NodeHandle nh_priv("~");
 
-  ekf.run();
+  RobotLocalization::RosEkf ekf(nh, nh_priv);
+  ekf.initialize();
+  ros::spin();
 
-  return 0;
+  return EXIT_SUCCESS;
```
Comment (ayrton04) [inline src/ekf_localization_node.cpp:48]: What defines this?

File: src/ekf_localization_nodelet.cpp [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+#include "robot_localization/ros_filter_types.h"
+
+#include <ros/ros.h>
+#include <nodelet/nodelet.h>
+
+namespace RobotLocalization
+{
+  class EkfNodelet : public nodelet::Nodelet
+  {
+  std::auto_ptr<RosEkf> ekf;
```
Comment (ayrton04) [inline src/ekf_localization_nodelet.cpp]: Even though this is supported in Kinetic, I'd like all of these changes to not require C++11, as it means I'll have to only allow this in Kinetic, and will have to maintain Indigo separately. If we have no choice, then we'll deal with it, but I'd prefer to support every ROS distribution that's currently not EOL, if possible.

File: src/ekf_localization_nodelet.cpp [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+  public:
+    virtual void onInit() {
```
Comment (ayrton04) [inline src/ekf_localization_nodelet.cpp]: No cuddled braces. Please see http://wiki.ros.org/CppStyleGuide. I realize the variables  don't adhere to standards, but I try to keep everything else that way.

File: src/ekf_localization_nodelet.cpp [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+      ros::NodeHandle nh      = getNodeHandle();
+      ros::NodeHandle nh_priv = getPrivateNodeHandle();
+
+      ekf.reset(new RosEkf(nh, nh_priv));
+      ekf->initialize();
+    }
+  };
+}
+
+#include <pluginlib/class_list_macros.h>
```
Comment (ayrton04) [inline src/ekf_localization_nodelet.cpp]: Can this include go at the top?

File: src/navsat_transform_nodelet.cpp [hunk 9]
```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2014, 2015, 2016, Charles River Analytics, Inc.
```
Comment (ayrton04) [inline src/navsat_transform_nodelet.cpp]: Don't need to make this CRA's copyright.

**Review by ayrton04 — CHANGES_REQUESTED** [review]
I'll likely have more, and I'm guessing some of these changes are in the other PR, but there are a few things to change.

File: src/ekf_localization_node.cpp (hunk 5 above)
Comment (SGSSGene) [reply src/ekf_localization_node.cpp:48]: This is standard c/c++: http://en.cppreference.com/w/cpp/utility/program/EXIT_status

File: src/navsat_transform_nodelet.cpp (hunk 9 above)
Comment (SGSSGene) [reply src/navsat_transform_nodelet.cpp]: Good point :-). This will take some time to clarify. Not sure which license I have to apply.

Comment (ayrton04) [reply src/navsat_transform_nodelet.cpp]: The package license is BSD, so we don't need to change that verbiage, but the copyright holder can be you or your organization. Note that I am not a copyright lawyer, but I'm pretty sure that's how it works. :)

File: src/ekf_localization_node.cpp (hunk 5 above)
Comment (ayrton04) [reply src/ekf_localization_node.cpp:48]: Right, but it needs you to `#include <cstdlib>`.

File: src/ekf_localization_nodelet.cpp (hunk 6 above)
Comment (SGSSGene) [reply src/ekf_localization_nodelet.cpp]: std::auto_ptr is c++03. It is not my favorite feature, but seemed better then a naked pointer.
In c++11 you would use std::unique_ptr instead.

**Comment by bochen87** [comment]
@ayrton04 Any reason this is not in kinetic?

**Comment by ayrton04** [comment]
It represents a pretty major change to the package, and was added well into Kinetic's lifecycle. I don't strictly adhere to the standards that OSRF suggests for managing changes within and across ROS distros, but this certainly seemed like a significant enough change that releasing it in Kinetic seemed questionable.

**Comment by bochen87** [comment]
ok, thanks for your explanation. Was just wondering, as it does seem to be a very useful feature!

**Comment by ayrton04** [comment]
If you don't mind working from source, the Lunar and Melodic branches will build just fine in Kinetic.

**Comment by bochen87** [comment]
ok, thanks for the hint!

**Pull request merged**